prog/kubernetes/provision_kubernetes_node.rb

```diff
@@ -0,0 +1,135 @@
+  sudo apt update
+  sudo apt install -y containerd
+  sudo mkdir -p /etc/containerd
+  sudo touch /etc/containerd/config.toml
+  containerd config default | sudo tee /etc/containerd/config.toml
+  sudo sed -i 's/SystemdCgroup = false/SystemdCgroup = true/' /etc/containerd/config.toml
+  sudo systemctl restart containerd
+  curl -fsSL https://pkgs.k8s.io/core:/stable:/#{kubernetes_cluster.kubernetes_version}/deb/Release.key | sudo gpg --dearmor -o /etc/apt/keyrings/kubernetes-apt-keyring.gpg
+  echo 'deb [signed-by=/etc/apt/keyrings/kubernetes-apt-keyring.gpg] https://pkgs.k8s.io/core:/stable:/#{kubernetes_cluster.kubernetes_version}/deb/ /' | sudo tee /etc/apt/sources.list.d/kubernetes.list
+  sudo apt update
+  sudo apt install -y kubelet kubeadm kubectl
+  sudo apt-mark hold kubelet kubeadm kubectl
+  sudo systemctl enable --now kubelet
+BASH
+
+    hop_bootstrap_rhizome
+  end
+
+  label def bootstrap_rhizome
+    bud Prog::BootstrapRhizome, {"target_folder" => "kubernetes", "subject_id" => vm.id, "user" => "ubi"}
+
+    hop_wait_bootstrap_rhizome
+  end
+
+  label def wait_bootstrap_rhizome
+    reap
+    hop_assign_role if leaf?
+    donate
+  end
+
+  label def assign_role
+    write_hosts_file_if_needed
+
+    hop_init_cluster if kubernetes_cluster.vms.count == 1
+
+    hop_join_control_plane
+  end
+
+  label def init_cluster
+    case vm.sshable.cmd("common/bin/daemonizer --check init_kubernetes_cluster")
+    when "Succeeded"
+      pop vm_id: vm.id
+    when "NotStarted"
+      ps = kubernetes_cluster.private_subnet
+      vm.sshable.cmd("common/bin/daemonizer 'sudo /home/ubi/kubernetes/bin/init-cluster #{kubernetes_cluster.name} #{kubernetes_cluster.endpoint} 443 #{ps.net4} #{ps.net6} #{vm.nics.first.private_ipv4}' init_kubernetes_cluster")
+      nap 30
+    when "InProgress"
+      nap 10
+    when "Failed"
+      puts "INIT CLUSTER FAILED"
+      nap 65536
+      # TODO: register deadline
+      #   fail "could not init control-plane cluster"
+    end
+
+    nap 65536
+  end
+
+  label def join_control_plane
+    case vm.sshable.cmd("common/bin/daemonizer --check join_control_plane")
+    when "Succeeded"
+      pop vm_id: vm.id
+    when "NotStarted"
+      join_token = kubernetes_cluster.vms.first.sshable.cmd("sudo kubeadm token create --ttl 24h --usages signing,authentication").tr("\n", "")
+      certificate_key = kubernetes_cluster.vms.first.sshable.cmd("sudo kubeadm init phase upload-certs --upload-certs")[/certificate key:\n(.*)/, 1]
+      discovery_token_ca_cert_hash = kubernetes_cluster.vms.first.sshable.cmd("sudo kubeadm token create --print-join-command")[/discovery-token-ca-cert-hash (.*)/, 1]
+
+      vm.sshable.cmd("common/bin/daemonizer 'sudo kubernetes/bin/join-control-plane-node #{kubernetes_cluster.endpoint}:443 #{join_token} #{discovery_token_ca_cert_hash} #{certificate_key}' join_control_plane")
```

Comment:
Should we call `.shellescape` for these values?

Reply:
Hmm this is also in a quote, so that means there's another layer of shell, which, fair question, should we...not do that? and rely on argv? is sudo even necessary given daemonizer implies root?

Reply:
I didn't understand this part:

> ... should we...not do that? and rely on argv?

Do you suggest pushing the info as a json in stdin for example? If not, an example would be awesome :)

Reply:
well, it was more like...do you need to quote it, or do you need only to terminate the command's option parsing? 

e.g. 

```
common/bin/daemonizer join_control_plane -- kubernetes/bin/join-control-plane-node #{kubernetes_cluster.endpoint}:443 #{join_token} #{discovery_token_ca_cert_hash} #{certificate_key}
```

....I see the problem now, with `join_control_plane` as a positional argument, it's plausible daemonizer is better to be changed, to have a v2, to avoid nested argument parsing when unnecessary. Also probably to avoid writing to /var/log and instead rely on the systemd journal.

Reply:
okay, I'm going to fool around with daemonizer, watch this space

Reply:
okay, I think I can improve this a bit. We also have some patches in this patch to daemonzer that may be a bit too disruptive to do in-situ, we may need to make daemonizer2 and lump together a few simplifications.

Reply:
>We also have some patches in this patch to daemonzer that may be a bit too disruptive to do in-situ, we may need to make daemonizer2 and lump together a few simplifications.

For clarification, if you're talking about [this](https://github.com/ubicloud/ubicloud/pull/2545/files#diff-5935f6c67f806cf538ca456dd1bf15e6a9f4be25b8e7498d3df7d28b4cf94124R68), I'm not planning to bring the daemonizer changes here.

Reply:
so, what's the recommended change for this now? shellescape?